### DISCUSSION TITLE
Investigate support for CPython 3.11

### DISCUSSION THREAD

**Discussion by antocuni** [pyscript/pyscript#1192] [Proposals]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other feature requests and couldn't find a duplicate (including also the ``type-feature`` tag)
- [X] I confirmed that it's not related to another project area (see the above section)

### What is the idea?

Currently, we are using a version of Pyodide which is based on CPython 3.10.
We would like to investigate using CPython 3.11, which is currently in beta.

CPython 3.11 should provide several advantages:
- it will have native support for WASM
- it should provide a much faster loading time.

An online demo of CPython 3.11 on WASM is available here: https://repl.ethanhs.me/

The easiest way to achieve this is to start a pyodide branch where to build pyodide with CPython 3.11.
Another possibility to explore is to use the vanilla build of CPython: this might be an interesting experiment, although it would probably lose the possibility to interact with the JS world which are offered by pyodide.

**Comment by rth** [comment]
It's worth exploring but I'm not convinced it would be radically faster to load. One reason Ethan's and Christian's REPL are faster to load is because they strip a lot things from CPython and do optimizations that work fine for a standalone Python REPL but that are more difficult if one wants to support a large number of Python packages. To give an example, Pyodide statically links  sqlite3 into the main module which increases its size. So there is some  features/size compromise, and a lot of these things are orthogonal to the Python version. But we can certainly re-evaluate some of these choices in Pyodide as well. 

For the case of sqlite3 we should probably switch to dynamic linking and loading it as a separate package, but then it means you have an extra dependency sqlite3 and so the assumption that the standard library just works, is no longer valid and one needs to add this dependency to any package that might potentially use it.

Also, note that a lot of packages currently don't yet officially support Python 3.11, so switching before the official release date, means dealing with those package issues that get typically ironed out a few months after the release in the package ecosystem. Probably not many issues as most things are forward compatible, but it would still add extra bug reports we don't necessarily want to deal with because they are upstream issues.

**Comment by antocuni** [comment]
> It's worth exploring but I'm not convinced it would be radically faster to load. One reason Ethan's and Christian's REPL are faster to load is because they strip a lot things from CPython and do optimizations that work fine for a standalone Python REPL but that are more difficult if one wants to support a large number of Python packages. 

ah that's good to know. I had read somewhere that they did some optimization specifically for the startup time on WASM, but I might be wrong and I certainly didn't do any measurement.
Maybe @tiran knows more?

> To give an example, Pyodide statically links sqlite3 into the main module which increases its size. So there is some features/size compromise, and a lot of these things are orthogonal to the Python version. But we can certainly re-evaluate some of these choices in Pyodide as well.

I understand why it makes sense for pyodide to build a distribute a "batteries included" default, but I am wondering whether it might be a good idea for pyscript to provide a stripped down version which is fast to load by default, and give the user an option to pay the extra penalty only if it's needed.

> For the case of sqlite3 we should probably switch to dynamic linking and loading it as a separate package, but then it means you have an extra dependency sqlite3 and so the assumption that the standard library just works, is no longer valid and one needs to add this dependency to any package that might potentially use it.

Ah I see. IUC this means that if I want to load an external package which assumes that `sqlite3` is available, I would need to call `loadPackage('sqlite3')` before installing/importing it?
Eventually, we might need some sort of dependency resolver so that the user can just list what he/she needs and let pyscript to do the rest.


> Also, note that a lot of packages currently don't yet officially support Python 3.11, so switching before the official release date, means dealing with those package issues that get typically ironed out a few months after the release in the package ecosystem. Probably not many issues as most things are forward compatible, but it would still add extra bug reports we don't necessarily want to deal with because they are upstream issues.

Yes, I'm not proposing to switch to 3.11 now, I just wanted to experiment with it and be "more ready" to transition once the rest of ecosystem is ready.

**Comment by tiran** [comment]
Python 3.11 has several core modules deep-frozen into the binary. Our build also ship pre-compiled ``pyc`` files instead of plain Python files. That helps with startup time, too.

* deep frozen core modules reduce startup time. Python 3.11 deep freezes a dozen modules like abc, io, os, site, and more https://github.com/faster-cpython/ideas/issues/84
* our Emscripten browser build use [PyZipfile](https://docs.python.org/3/library/zipfile.html#pyzipfile-objects) to ship pre-compiled ``pyc`` files instead of plain ``py`` files. This avoids initial cost for parsing and byte compiling Python code to byte code.
* the build on Ethan's website has no support for dynamic shared extensions. It has ``dlopen`` disabled, does not use position independent code (PIC), and does not permit side modules. MAIN_MODULE / SIDE_MODULE comes with some overhead.
* we ship less code in ``python.wasm``. AFAIK Pyodide adds ``libfreetype`` and several other big static libraries for 3rd party code.

**Comment by tiran** [comment]
Updated demo of 3.11.0b4 with dynamic linking enabled: https://cheimes.fedorapeople.org/python-wasm/

**Comment by antocuni** [comment]
> deep frozen core modules reduce startup time. Python 3.11 deep freezes a dozen modules like abc, io, os, site, and more https://github.com/faster-cpython/ideas/issues/84

this is interesting. I was already thinking of something like that (unsurprising, since we have been doing it for years in RPython) so I'm glad it has already been done :).
I wonder whether we could compile these python-modules-as-c-static-data into a WASM file and distribute/load it on demand. I guess it would need at least some kind of relocation logic, but it might be something worth exploring in the future.

**Comment by tiran** [comment]
FYI, wasm32-emscripten has reached PEP 11 support tier 3, https://github.com/python/steering-council/issues/131

**Comment by antocuni** [comment]
> FYI, wasm32-emscripten has reached PEP 11 support tier 3, https://github.com/python/steering-council/issues/131

woo, that's a wonderful news! Congrats!

**Comment by tiran** [comment]
@rth @rth @antocuni what is the progress of your investigation? Python 3.11rc1 will be released in a week from now. I can still address minor problems in 3.11 until mid next week. Once rc1 is out, the window of opportunity is closed.

**Comment by antocuni** [comment]
@tiran I got busy with other things so I haven't started yet :(

**Comment by rth** [comment]
In Pyodide missed the rc1 window, but hopefully, we would be able to update for the next release, now that 0.21 is out. Opened https://github.com/pyodide/pyodide/issues/3051 about it as well.
